From 186ecae08f4602896bf18a64c0aa2d83fa69849d Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Wed, 10 Jul 2024 09:55:44 +0300
Subject: [PATCH] [FSW-2704] fs1: image prefetch for ppboot

This patch adds flow to prefetch ppboot image
before manifest validation. This api will try
primary manifest first and if failed will try
secondary manifest for ppboot image. To make
this api non-blocking only first dma chunk is
initialized.

Change-Id: Ibef764322e86cd5bbea5a66a93ec445fe5a8d2ce
---

diff --git a/app/boot_rom.c b/app/boot_rom.c
index c4f42c8..02eb3ef 100644
--- a/app/boot_rom.c
+++ b/app/boot_rom.c
@@ -7,6 +7,7 @@
 #include <boot_rom.h>
 #include <security.h>
 #include <boot_flow.h>
+#include <recovery.h>
 
 const int fs1_die_type = DIE_TYPE;
 
@@ -37,7 +38,8 @@
 
 	boot_rom_tpm_spi_setup();
 
-	boot_rom_prefetch_img(HL_FW_IMG_PPBOOT);
+	if (boot_rom_boot_flow_prefetch_img(HL_FW_IMG_PPBOOT) != BOOT_ROM_SUCCESS)
+		execute_recovery();
 
 	boot_rom_fake_fuse();
 
diff --git a/common/boot_flow.c b/common/boot_flow.c
index 149eca9..91b70df 100644
--- a/common/boot_flow.c
+++ b/common/boot_flow.c
@@ -5,19 +5,68 @@
  */
 
 #include <boot_rom.h>
+#include <hl_boot_desc.h>
+#include <boot_desc.h>
+#include <progress_bits.h>
+#include <dma.h>
 
-uint32_t boot_rom_prefetch_img(__unused enum hl_fw_img_type img_type)
+static enum hl_fw_manifest_type primary_manifest_idx = HL_FW_SPI_MANIFEST_0;
+
+static uint32_t boot_flow_prefetch_img_sub(enum hl_fw_manifest_type manifest_idx,
+					enum hl_fw_img_type img_type, struct boot_desc *desc,
+					enum boot_rom_progress img_sts_base)
 {
-	/* start the prefetch of ppboot or recovery images
-	 * In either case there can be multiple dmas involved
-	 * Maintain a state machine to do multiple dmas
-	 * It needs to decide which manifest it needs to do prefetch
-	 * from. If manifest crc or validation is invalid then it
-	 * needs to make sure to fetch image from other manifest.
-	 * Once the dma starts with one manifest, another function will
-	 * be implemented to complete it based on state machine.
-	 * return the manifest index from which it has started prefetch.
+	int err;
+
+	if (!boot_rom_boot_desc_init_done(manifest_idx))
+		return BOOT_ROM_FAIL;
+
+	/* get image descriptor from the manifest */
+	if (boot_desc_get_img_desc(manifest_idx, img_type, desc) != BOOT_ROM_SUCCESS)
+		return BOOT_ROM_FAIL;
+
+	/* Validate Image descriptor start address and other ranges */
+	if (boot_rom_verify_img_desc_address_range(desc, img_sts_base, img_type)
+	    != BOOT_ROM_SUCCESS)
+		return BOOT_ROM_FAIL;
+
+	err = boot_rom_spi_dma(desc->img_size, desc->spi_offset, desc->loadaddr,
+			       false, false, 0);
+	if (boot_rom_process_dma_err(err, img_sts_base))
+		return BOOT_ROM_FAIL;
+
+	return BOOT_ROM_SUCCESS;
+}
+
+uint32_t boot_rom_boot_flow_prefetch_img(enum hl_fw_img_type img_type)
+{
+	struct boot_desc desc = {0};
+	/* TODO: Image status base is complicated. Handle based on img_type */
+	enum boot_rom_progress img_sts_base = PROGRESS_BRINGUP_PCI_PRI;
+	enum hl_fw_manifest_type curr_manifest_idx;
+
+	if (!boot_rom_in_master_die())
+		return BOOT_ROM_SUCCESS;
+
+	/*
+	 * Check if manifest switch is set for fw update.
+	 * All manifest switches should be tracked through
+	 * single source file boot_flow.c
 	 */
+	/* TODO: Add check if secondary image is supported */
+	if (!!boot_rom_global.fw_upd_switch_manifest) {
+		primary_manifest_idx = HL_FW_SPI_MANIFEST_1;
+		img_sts_base = PROGRESS_BRINGUP_PCI_SEC;
+	}
+
+	curr_manifest_idx = primary_manifest_idx;
+
+	if (boot_flow_prefetch_img_sub(curr_manifest_idx, img_type, &desc, img_sts_base) !=
+	    BOOT_ROM_SUCCESS) {
+		/* TODO: Add check if secondary image is supported */
+		curr_manifest_idx = (primary_manifest_idx + 1) % HL_FW_SPI_MANIFEST_MAX_NUM;
+		return boot_flow_prefetch_img_sub(curr_manifest_idx, img_type, &desc, img_sts_base);
+	}
 
 	return BOOT_ROM_SUCCESS;
 }
diff --git a/common/boot_flow.h b/common/boot_flow.h
index 0612c5c..fc2e128 100644
--- a/common/boot_flow.h
+++ b/common/boot_flow.h
@@ -6,6 +6,6 @@
 #ifndef _BOOT_FLOW_H_
 #define _BOOT_FLOW_H_
 
-uint32_t boot_rom_prefetch_img(enum hl_fw_img_type img_type);
+uint32_t boot_rom_boot_flow_prefetch_img(enum hl_fw_img_type img_type);
 
 #endif /* _BOOT_FLOW_H_ */
